api/api_test.go

```diff
@@ -0,0 +1,446 @@
+	"github.com/qri-io/qri/repo/test"
+	regmock "github.com/qri-io/registry/regserver/mock"
+)
+
+func init() {
+	abide.SnapshotsDir = "testdata"
+}
+
+func newTestRepo(t *testing.T) (r repo.Repo, teardown func()) {
+	var err error
+	if err = confirmQriNotRunning(); err != nil {
+		t.Fatalf(err.Error())
+	}
+
+	// bump up log level to keep test output clean
+	golog.SetLogLevel("qriapi", "error")
+
+	// use a test registry server & client & client
+	rc, registryServer := regmock.NewMockServer()
+	// we need to artificially specify the timestamp
+	// we use the dsfs.Timestamp func variable to override
+	// the actual time
+	prevTs := dsfs.Timestamp
+	dsfs.Timestamp = func() time.Time { return time.Date(2001, 01, 01, 01, 01, 01, 01, time.UTC) }
+
+	if r, err = test.NewTestRepo(rc); err != nil {
+		t.Fatalf("error allocating test repo: %s", err.Error())
+	}
+
+	lib.Config = config.DefaultConfig()
+	lib.Config.Profile = test.ProfileConfig()
+	lib.Config.Registry.Location = registryServer.URL
+	prevSaveConfig := lib.SaveConfig
+	lib.SaveConfig = func() error {
+		p, err := profile.NewProfile(lib.Config.Profile)
+		if err != nil {
+			return err
+		}
+
+		r.SetProfile(p)
+		return err
+	}
+
+	teardown = func() {
+		golog.SetLogLevel("qriapi", "info")
+		lib.SaveConfig = prevSaveConfig
+		dsfs.Timestamp = prevTs
+	}
+
+	return
+}
+
+type handlerTestCase struct {
+	method, endpoint string
+	body             []byte
+}
+
+// runHandlerTestCases executes a slice of handlerTestCase against a handler
+func runHandlerTestCases(t *testing.T, name string, h http.HandlerFunc, cases []handlerTestCase) {
+	for i, c := range cases {
+		name := fmt.Sprintf("%s %s case %d: %s %s", t.Name(), name, i, c.method, c.endpoint)
+		req := httptest.NewRequest(c.method, c.endpoint, bytes.NewBuffer(c.body))
+		// TODO - make this settable with some sort of test case interface
+		req.Header.Set("Content-Type", "application/json")
+		w := httptest.NewRecorder()
+
+		h(w, req)
+
+		res := w.Result()
+		abide.AssertHTTPResponse(t, name, res)
+	}
+}
+
+// mustFile reads file bytes, calling t.Fatalf if the file doesn't exist
+func mustFile(t *testing.T, filename string) []byte {
+	data, err := ioutil.ReadFile(filename)
+	if err != nil {
+		t.Fatalf("error opening test file: %s: %s", filename, err.Error())
+	}
+	return data
+}
+
+func confirmQriNotRunning() error {
+	l, err := net.Listen("tcp", fmt.Sprintf(":%d", config.DefaultAPIPort))
+	if err != nil {
+		return fmt.Errorf("it looks like a qri server is already running on port %d, please close before running tests", config.DefaultAPIPort)
+	}
+
+	l.Close()
+	return nil
+}
+
+func TestServerRoutes(t *testing.T) {
+	r, teardown := newTestRepo(t)
+	defer teardown()
+
+	h := NewRootHandler(NewDatasetHandlers(r, false), NewPeerHandlers(r, nil, false))
+	rootCases := []handlerTestCase{
+		{"OPTIONS", "/", nil},
+		{"GET", "/", nil},
+	}
+	runHandlerTestCases(t, "root", h.Handler, rootCases)
+
+	healthCheckCases := []handlerTestCase{
```

Comment:
These two arrays are the same, will they always be the same? How about just using the same one twice.

Reply:
They should diverge over time, just wanted to get them in place as such for basic coverage.